platform/entanglement/src/services/CopyService.js

```diff
@@ -54,45 +57,156 @@ define(
             );
         };
 
-        CopyService.prototype.perform = function (domainObject, parent) {
-            var model = JSON.parse(JSON.stringify(domainObject.getModel())),
+        /**
+         * Will build a graph of an object and all of its child objects in
+         * memory
+         * @param domainObject The original object to be copied
+         * @param parent The parent of the original object to be copied
+         * @returns {Promise} resolved with an array of clones of the models
+         * of the object tree being copied. Copying is done in a bottom-up
+         * fashion, so that the last member in the array is a clone of the model
+         * object being copied. The clones are all full composed with
+         * references to their own children.
+         */
+        CopyService.prototype.buildCopyPlan = function(domainObject, parent, progress) {
+            var clones = [],
                 $q = this.$q,
                 self = this;
+            
+            function makeClone(object) {
+                return JSON.parse(JSON.stringify(object));
+            }
 
-            // Wrapper for the recursive step
-            function duplicateObject(domainObject, parent) {
-                return self.perform(domainObject, parent);
+            /**
+             * A recursive function that will perform a bottom-up copy of
+             * the object tree with originalObject at the root. Recurses to
+             * the farthest leaf, then works its way back up again,
+             * cloning objects, and composing them with their child clones
+             * as it goes
+             * @param originalObject
+             * @param originalParent
+             * @returns {*}
+             */
+            function copy(originalObject, originalParent) {
+                //Make a clone of the model of the object to be copied
+                var modelClone = {
+                        id: uuid(),
+                        model: makeClone(originalObject.getModel()),
+                        persistenceSpace: originalParent.getCapability('persistence')
+                    };
+                delete modelClone.model.composition;
+                delete modelClone.model.persisted;
+                delete modelClone.model.modified;
+                return $q.when(originalObject.useCapability('composition')).then(function(composees){
+                    progress({phase: "preparing"});
+                    return (composees || []).reduce(function(promise, composee){
+                            //If the object is composed of other
```

Comment:
Oh man, nesting gets way too deep here, so I welcome any suggestions for improving readability. I'd like to revisit this function post-#215 actually and see if it can't be refactored to be a little cleaner.